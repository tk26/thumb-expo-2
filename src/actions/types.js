export const FOLLOW_USER = 'FOLLOW_USER';
export const UNFOLLOW_USER = 'UNFOLLOW_USER';

//Auth
export const EMAIL_CHANGED = 'EMAIL_CHANGED';
export const PASSWORD_CHANGED = 'PASSWORD_CHANGED';
export const LOGIN_USER = 'LOGIN_USER';
export const LOGIN_USER_FAILED = 'LOGIN_USER_FAILED';
export const LOGIN_USER_SUCCESS = 'LOGIN_USER_SUCCESS';
export const LOGIN_USER_AUTH_FAILED = 'LOGIN_USER_AUTH_FAILED';
export const LOGIN_UNVERIFIED_USER_FAILED = 'LOGIN_UNVERIFIED_USER_FAILED';
export const LOGOUT_USER = 'LOGOUT_USER';

//Profile
export const PROFILE_UPDATED = 'PROFILE_UPDATED';
export const PROFILE_LOGOUT = 'PROFILE_LOGOUT';

//Sign-Up
export const SIGNUP_UPDATE = 'SIGNUP_UPDATED';
export const SIGNUP_SUBMIT_STEP = 'SIGNUP_SUBMIT_STEP';
export const SIGNUP_STEP1_SUCCESS = 'SIGNUP_STEP1_SUCCESS';
export const SIGNUP_STEP1_ERROR = 'SIGNUP_STEP1_ERROR';
export const SIGNUP_STEP2_SUCCESS = 'SIGNUP_STEP2_SUCCESS';
export const SIGNUP_STEP2_ERROR = 'SIGNUP_STEP2_ERROR';
export const SIGNUP_STEP3_SUCCESS = 'SIGNUP_STEP3_SUCCESS';
export const SIGNUP_STEP3_ERROR = 'SIGNUP_STEP3_ERROR';
export const SIGNUP_STEP4_SUCCESS = 'SIGNUP_STEP4_SUCCESS';
export const SIGNUP_STEP4_ERROR = 'SIGNUP_STEP4_ERROR';