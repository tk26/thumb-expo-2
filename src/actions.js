export const FOLLOW_USER = 'FOLLOW_USER';

export const UNFOLLOW_USER = 'UNFOLLOW_USER';

// Imp: actions must have a type

export function followUser(username) {
  return { type: FOLLOW_USER, username };
}

export function unfollowUser(username) {
  return { type: UNFOLLOW_USER, username };
}