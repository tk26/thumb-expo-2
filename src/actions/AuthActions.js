import { AuthService } from '../services' ;
import { EMAIL_CHANGED, 
        PASSWORD_CHANGED, 
        LOGIN_USER,
        LOGIN_USER_SUCCESS,
        LOGIN_USER_FAILED,
        LOGIN_USER_AUTH_FAILED,
        LOGIN_UNVERIFIED_USER_FAILED,
        LOGOUT_USER} from './types';


// Imp: actions must have a type
export function emailChanged(email) {
    return { type: EMAIL_CHANGED, email };
}
  
export function passwordChanged(password) {
    return { type: PASSWORD_CHANGED, password };
}

export function loginUser({ email, password }){
    return async(dispatch) => {
        dispatch({ type: LOGIN_USER });
        const error = AuthService.validateEmailAndPassword(email, password);
        if (error !== ''){
            loginUserFailWithError(dispatch, error);
            return;
        }
        try {
            let response = await AuthService.login(email, password);
            switch (response.status){
                case 403:
                    return loginUnverifiedUserFail(dispatch);
                case 400:
                    return loginUserAuthFail(dispatch);
                case 200:
                    return loginSuccess(dispatch, response);
                default:
                    return loginUserFail(dispatch);            
            }
        }
        catch(error){
            loginUserFail(dispatch);
        }
    };
}

export function logoutUser(){
    AuthService.logout();
    return { type: LOGOUT_USER };
}

export function loginFailed(){
    return { type: LOGIN_USER_FAILED };
}

const loginUserFailWithError = (dispatch, error) => {
    dispatch({ type: LOGIN_USER_FAILED, error: error });
};

const loginUserAuthFail = (dispatch) => {
    dispatch({ type: LOGIN_USER_AUTH_FAILED });
};

const loginUserFail = (dispatch) => {
    dispatch({ type: LOGIN_USER_FAILED });
};

const loginUnverifiedUserFail = (dispatch) => {
    dispatch({ type: LOGIN_UNVERIFIED_USER_FAILED });
};

const loginSuccess = (dispatch, rawResponse) => {
    rawResponse.json()
        .then((response) => {
            const auth_token = response.token;
            AuthService.setAuthToken(auth_token);
            // Save user details
            let profile = {
                firstName: response.firstName,
                lastName: response.lastName,
                school: response.school,
                username: response.username,
                profilePicture: response.profilePicture,
                birthday: response.birthday,
                bio: response.bio
            };
            global.firstName = profile.firstName;
            global.profilePicture = profile.profilePicture;
            dispatch({type: LOGIN_USER_SUCCESS, token: auth_token});
            dispatch({type: PROFILE_UPDATED, profile: profile});
        })
        .catch((dispatch) => loginUserFail(dispatch));
}