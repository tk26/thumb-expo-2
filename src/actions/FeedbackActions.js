import { FeedbackService } from '../services';
import { FEEDBACK_UPDATE,
    FEEDBACK_SUBMIT_INIT,
    FEEDBACK_SUBMIT_SUCCESS,
    FEEDBACK_SUBMIT_ERROR
} from './types';
import { getApiUrl } from '../helper';
import { fetchWithTokenHandler } from '../services/fetchPlus';

export const unexpectedException = "Some error occured. Please try again. If problem persists, " +
    "please let us know at support@thumbtravel.com";

export const feedbackUpdate = ({prop, value}) => {
    return {
        type: FEEDBACK_UPDATE,
        payload: {prop, value}
    }
}

export const submitFeedback = ({feedbackType, feedbackDescription}) => {
    return async(dispatch) => {
        dispatch({ type: FEEDBACK_SUBMIT_INIT });
        if (feedbackDescription.length < 1) {
            return dispatch ({ type: FEEDBACK_SUBMIT_ERROR, error: 'Invalid feedback details' });
        }
        try {
          let response = await fetchWithTokenHandler(getApiUrl() + '/feedback/submit', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
              "feedbackType": feedbackType,
              "feedbackDescription": feedbackDescription
            })
          }, dispatch);

          switch(response.status){
              case 200:
                  return dispatch ({ type: FEEDBACK_SUBMIT_SUCCESS });
              case 400:
                  return dispatch ({ type: FEEDBACK_SUBMIT_ERROR, error: 'Invalid user details' });
              default:
                  return dispatch ({ type: FEEDBACK_SUBMIT_ERROR, error: unexpectedException });
              }
        } catch(error){
          return dispatch ({ type: FEEDBACK_SUBMIT_ERROR, error: unexpectedException });
        }
    }
}
