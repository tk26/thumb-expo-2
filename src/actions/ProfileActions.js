import { PROFILE_LOGOUT } from './types';

export function clearProfile(){
    return { type: PROFILE_LOGOUT };
}
