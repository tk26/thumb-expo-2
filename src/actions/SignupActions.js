import { AuthService } from '../services';
import { SIGNUP_UPDATE,
  SIGNUP_SUBMIT_STEP,
  SIGNUP_STEP1_SUCCESS,
  SIGNUP_STEP1_ERROR,
  SIGNUP_STEP2_SUCCESS,
  SIGNUP_STEP2_ERROR,
  SIGNUP_STEP3_SUCCESS,
  SIGNUP_STEP3_ERROR
} from './types';


export const signupUpdate = ({prop, value}) => {
  return {
    type: SIGNUP_UPDATE,
    payload: {prop, value}
  }
}

export const submitStep1 = ({firstName, lastName, username}) => {
  return async(dispatch) => {
    dispatch({ type: SIGNUP_SUBMIT_STEP });
    const usernameRegex = /^[a-z0-9._]{3,30}$/;
    const step = 1;

    if (firstName.length < 1) {
      return stepFailed(dispatch, step, "First Name cannot be empty");
    }

    if (lastName.length < 1) {
      return stepFailed(dispatch, step, "Last Name cannot be empty");
    }

    if(!usernameRegex.test(username)){
      return stepFailed(dispatch, step, "Username should be between 3 to 30 characters " +
                      "and can only contain numbers, letters, periods and underscores");
    }

    const unexpectedException = "Some error occured. Please try again. If problem persists, " +
      "please let us know at support@thumbtravel.com";

    try {
      let response = await AuthService.validateUsername(username);
      switch (response.status){
        case 422:
            return stepFailed(dispatch, step, 'Invalid username');
        case 409:
            return stepFailed(dispatch, step, 'Duplicate username');
        case 200:
            return stepSucceeded(dispatch, step, response);
        default:
            return stepFailed(dispatch, step, unexpectedException);
      }
    }
    catch(error) {
      stepFailed(dispatch, step, unexpectedException);
    }
  }
}

export const submitStep2 = ({password, confirmPassword}) => {
  const passwordRegex = new RegExp("^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#\$%\^&\*])(?=.{8,})");
  const step = 2;
  return (dispatch) => {
    if (password.length < 8 || password.length > 30) {
      return stepFailed(dispatch, step, "Password should be between 8 to 30 characters" );
    }
    if (!passwordRegex.test(password)) {
      const error = "Password should be a combinaton of upper and lowercase letters, " +
          "a number and a special character";
      return stepFailed(dispatch, step, error);
    }
    if (password !== confirmPassword) {
      return stepFailed(dispatch, step, "Password and Confirm Password do not match");
    }
    return stepSucceeded(dispatch, step);
  }
}

const stepSucceeded = (dispatch, step) => {
  let type;
  console.log(step);
  switch (step){
    case 1:
      type = SIGNUP_STEP1_SUCCESS;
      break;
    case 2:
      type = SIGNUP_STEP2_SUCCESS;
      break;
    default:
      type = SIGNUP_STEP3_SUCCESS;
      break;
  }

  dispatch ({
    type: type
  });
}

const stepFailed = (dispatch, step, error) => {
  let type;
  switch (step){
    case 1:
      type = SIGNUP_STEP1_ERROR;
    case 2:
      type = SIGNUP_STEP2_ERROR;
    default:
      type = SIGNUP_STEP3_ERROR;
  }
  dispatch ({
    type: type,
    error
  });
}
