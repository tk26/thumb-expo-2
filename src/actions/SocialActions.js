import { FOLLOW_USER, UNFOLLOW_USER } from './types';

// Imp: actions must have a type
export function followUser(username) {
    return { type: FOLLOW_USER, username };
  }
  
  export function unfollowUser(username) {
    return { type: UNFOLLOW_USER, username };
  }