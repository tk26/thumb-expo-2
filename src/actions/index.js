export * from './SocialActions';
export * from './AuthActions';
export * from './ProfileActions';
export * from './SignupActions';