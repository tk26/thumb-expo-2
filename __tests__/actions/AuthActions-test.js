import configureStore from 'redux-mock-store';

// Actions to be tested
import * as authActions from '../../src/actions/AuthActions';
import * as types from '../../src/actions/types';

const mockStore = configureStore();
const store = mockStore();

describe('Auth_Actions', () => {
    beforeEach(() => { // Runs before each test in the suite
      store.clearActions();
    });
    describe('email change', () => {
        test('Dispatches the correct action and payload', () => {
            const email = 'testemail@testing.edu';    
            store.dispatch(authActions.emailChanged(email));
            expect(store.getActions()).toMatchSnapshot();
        });
    });
    describe('password change', () => {
        test('Dispatches the correct action and payload', () => {
            const password = 'Password123!';  
            store.dispatch(authActions.passwordChanged(password));
            expect(store.getActions()).toMatchSnapshot();
        });
    });
});