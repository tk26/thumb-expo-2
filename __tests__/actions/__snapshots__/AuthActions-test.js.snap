// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth_Actions email change Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "email": "testemail@testing.edu",
    "type": "EMAIL_CHANGED",
  },
]
`;

exports[`Auth_Actions password change Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "password": "Password123!",
    "type": "PASSWORD_CHANGED",
  },
]
`;
